rust/lance/src/dataset/scanner.rs

```diff
@@ -1027,11 +1028,22 @@ impl Scanner {
         Ok(concat_batches(&schema, &batches)?)
     }
 
-    /// Scan and return the number of matching rows
-    #[instrument(skip_all)]
-    pub fn count_rows(&self) -> BoxFuture<Result<u64>> {
+    fn create_count_plan(&self) -> BoxFuture<Result<Arc<dyn ExecutionPlan>>> {
         // Future intentionally boxed here to avoid large futures on the stack
         async move {
+            if !self.projection_plan.physical_schema.fields.is_empty() {
+                return Err(Error::invalid_input(
+                    "count_rows should not be called on a plan selecting columns".to_string(),
+                    location!(),
+                ));
+            }
```

Comment:
I'm a little torn on this error.  Ideally, we would just silently blank out the projection plan and then create the count plan.  However, to do that we either have to clone the scan, which is a pretty big thing to be cloning, or modify the scanner, which would maybe not be what users would expect from `count_rows`.

For now, I want to get something out soon, so I'm just raising an error, with the assumption that `Scanner::count_rows` is a mostly internal method anyways (users should use `Dataset::count_rows`).

Reply:
I think you're right that is pretty internal. Plus easy to work around.